infra/Pos/Infra/Semaphore.hs

```diff
@@ -0,0 +1,80 @@
+-- | 'BlkSemaphore' primitive used for synchronization of threads in
+-- the application.
+
+module Pos.Infra.Semaphore
+       ( BlkSemaphore (..)
+       , withBlkSemaphore
+       , withBlkSemaphoreIgnoreTip
+       , withBlkSemaphore_
+       ) where
+
+import           Universum
+
+import           Control.Monad.Catch (MonadMask, bracketOnError)
+
+import           Pos.Core            (HeaderHash)
+import           Pos.Util.Util       (HasLens, lensOf)
+
+-- | A simple wrapper over 'MVar' which stores 'HeaderHash' (our
+-- current tip) and is taken whenever we want to update GState or
+-- other data dependent on GState.
+newtype BlkSemaphore = BlkSemaphore
+    { unBlkSemaphore :: MVar HeaderHash
+    }
+
+-- TODO: use 'withMVar' instead ('MonadThrow'-based).
```

Comment:
I think I will do it in this PR.
It's not used, because it's not defined 😉 
As I wrote below, there is `IO` version in `base` and `MonadBaseControl IO` version in `lifted-base`, but no `MonadMask` version.